docs/source/hsm.md

```diff
@@ -0,0 +1,161 @@
+# Using a Hardware Security Module (HSM)
+
+You can use a Hardware Security Module (HSM) to generate and store the private
+keys used by your Fabric nodes. An HSM protects your private keys and handles
+cryptographic operations, which allows your peers and ordering nodes to sign and
+endorse transactions without exposing their private keys. Currently, Fabric only
+supports the PKCS11 standard to communicate with an HSM.
+
+## Configuring an HSM
+
+To use an HSM with your Fabric node, you need to update the BCCSP (Crypto Service
+Provider) section of the node configuration file such as core.yaml or
+orderer.yaml. In BCCSP section, you need to select PKCS11 as the provider and
+provide the path to the PKCS11 library that you would like to use. You also need
+to provide the label and pin of the token that you created for your cryptographic
+operations. You can use one token to generate and store multiple keys.
+
+The prebuilt Hyperledger Fabric Docker images are not enabled to use PKCS11. If
+you are deploying Fabric using docker, you need to build your own images and
+enable PKCS11 using the following command:
+```
+make docker GO_TAGS=pkcs11
+```
+You also need to ensure that the PKCS11 library is available to be used by the
+node by installing it or mounting it inside the container.
+
+### Example
+
+The following example demonstrates how to configure a Fabric node to use an HSM.
+
+First, you will need to install an implementation of the PKCS11 interface. This
+example uses the [softhsm](https://github.com/opendnssec/SoftHSMv2) open source
+implementation. After downloading and configuring softhsm, you will need to set
+the SOFTHSM2_CONF environment variable to point to the softhsm2 configuration
+file.
+
+You can then use softhsm to create the token that will handle the cryptographic
+operations of your Fabric node inside an HSM slot. In this example, we create a
+token labelled "fabric" and set the pin to "71811222". After you have created
+the token, update the configuration file to use PKCS11 and your token as the
+crypto service provider. You can find an example BCCSP section below:
+
+```
+#############################################################################
+# BCCSP (BlockChain Crypto Service Provider) section is used to select which
+# crypto library implementation to use
+#############################################################################
+bccsp:
+  default: PKCS11
+  pkcs11:
+    Library: /etc/hyperledger/fabric/libsofthsm2.so
+    Pin: 71811222
+    Label: fabric
+    hash: SHA2
+    security: 256
+    Immutable: false
+```
```

Comment:
In v1.4.x, there is also properties for FileKeyStore and KeyStore. These were removed in v2.0 since they are not needed. But it is unclear if they have to be configured in v1.4.x... see:
https://github.com/hyperledger/fabric/blob/release-1.4/sampleconfig/core.yaml#L298-L299

Reply:
FileKeyStore and KeyStore is used for Software based BCCSP and hence does not play in PKCS11 configuration. For that reason , it should be mentioned here IMO.

Reply:
I assume you mean 'should not' be mentioned here.
So is the suggestion to simply leave these fields blank in core.yaml for people using HSM on 1.4?

Reply:
Confirmed offline with Ash that he meant `FileKeyStore and KeyStore is used for Software based BCCSP and hence does not play in PKCS11 configuration. For that reason , it should not be mentioned here IMO.`